src/object_ref.jl

```diff
@@ -1,12 +1,60 @@
 mutable struct ObjectRef
-    oid::ray_jll.ObjectIDAllocated
+    # oid::ray_jll.ObjectIDAllocated
+    oid_hex::String
     owner_address::Union{ray_jll.AddressAllocated,Nothing}
     serialized_object_status::String
+
+    function ObjectRef(oid_hex, owner_address, serialized_object_status;
+                       add_local_ref=true)
+        objref = new(oid_hex, owner_address, serialized_object_status)
+        if add_local_ref
+            worker = ray_jll.GetCoreWorker()
+            ray_jll.AddLocalReference(worker, objref.oid)
+        end
+        # finalizer(f, object) returns object
+        return finalizer(objref) do objref
+            errormonitor(@async finalize_object_ref(objref))
+            return nothing
+        end
+    end
+end
+
+function finalize_object_ref(obj::ObjectRef)
+    @debug "Removing local ref for ObjectID $(obj.oid_hex)"
+    worker = ray_jll.GetCoreWorker()
+    oid = ray_jll.FromHex(ray_jll.ObjectID, obj.oid_hex)
+    ray_jll.RemoveLocalReference(worker, oid)
+    return nothing
+end
+
+function Base.getproperty(x::ObjectRef, prop::Symbol)
+    return if prop == :oid
+        ray_jll.FromHex(ray_jll.ObjectID, x.oid_hex)
+    else
+        getfield(x, prop)
+    end
+end
+
+# in order to actually increment the local ref count appropriately when we
+# `deepcopy` an ObjectRef and setup the appropriate finalizer, this
+# specialization calls the constructor after deepcopying the fields.
+function Base.deepcopy_internal(x::ObjectRef, stackdict::IdDict)
+    fieldnames = Base.fieldnames(typeof(x))
+    fieldcopies = ntuple(length(fieldnames)) do i
+        @debug "deep copying x.$(fieldnames[i])"
+        fieldval = getfield(x, fieldnames[i])
+        return Base.deepcopy_internal(fieldval, stackdict)
+    end
+
+    xcp = ObjectRef(fieldcopies...; add_local_ref=true)
+    stackdict[x] = xcp
+
+    return xcp
 end
```

Comment:
So the `owner_address` and `serialized_object_status` fields only exist for delaying registration until after deserialization. You can just avoid populating those fields in `deepcopy` as they're only for internal use and we could probably simply this down to just dealing with the `ObjectID`.

Reply:
maybe we coudl, but I don't see the downside of honoring the spirit of deepcopy unless there's a very good reason not to.